librad/src/executor.rs

```diff
@@ -0,0 +1,333 @@
+        } else {
+            JoinHandle::Cancelled
+        }
+    }
+
+    pub fn spawn_blocking<F, T>(&self, f: F) -> JoinHandle<T>
+    where
+        F: FnOnce() -> T + Send + 'static,
+        T: Send + 'static,
+    {
+        if !self.detach.is_closed() {
+            let span = tracing::Span::current();
+            JoinHandle::Spawned(SpawnedTask {
+                _detach: self.detach.clone(),
+                task: Some(self.inner.spawn_blocking(move || {
+                    let _guard = span.enter();
+                    f()
+                })),
+            })
+        } else {
+            JoinHandle::Cancelled
+        }
+    }
+}
+
+impl From<tokio::runtime::Handle> for Spawn {
+    fn from(hdl: tokio::runtime::Handle) -> Self {
+        Self::new(hdl)
+    }
+}
+
+#[must_use = "spawned tasks must be awaited"]
+pub enum JoinHandle<T> {
+    Cancelled,
+    Spawned(SpawnedTask<T>),
+}
+
+impl<T> JoinHandle<T> {
+    pub fn abort(&self) {
+        if let Self::Spawned(task) = self {
+            task.abort()
+        }
+    }
+}
+
+impl JoinHandle<()> {
+    pub fn detach(self) {
+        if let Self::Spawned(task) = self {
+            task.detach()
+        }
+    }
+}
+
+impl<T> Future for JoinHandle<T> {
+    type Output = Result<T, JoinError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        match self.get_mut() {
+            JoinHandle::Cancelled => Poll::Ready(Err(JoinError::Cancelled)),
+            JoinHandle::Spawned(ref mut task) => Pin::new(task).poll(cx),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum JoinError {
+    #[error("task cancelled")]
+    Cancelled,
+    #[error("task panicked")]
+    Panicked(Box<dyn Any + Send + 'static>),
+}
+
+impl JoinError {
+    pub fn is_panic(&self) -> bool {
+        match self {
+            Self::Cancelled => false,
+            Self::Panicked(_) => true,
+        }
+    }
+
+    pub fn into_cancelled(self) -> Cancelled {
+        match self {
+            Self::Cancelled => Cancelled,
+            Self::Panicked(panik) => panic::resume_unwind(panik),
+        }
+    }
+}
+
+impl From<tokio::task::JoinError> for JoinError {
+    fn from(e: tokio::task::JoinError) -> Self {
+        if e.is_cancelled() {
+            Self::Cancelled
+        } else if e.is_panic() {
+            Self::Panicked(e.into_panic())
+        } else {
+            unreachable!("unexpected join error: {:?}", e)
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+#[error("spawned task cancelled")]
+pub struct Cancelled;
+
+struct Notify {
+    stop: oneshot::Receiver<()>,
+    done: Arc<Done>,
+}
+
+struct Done {
+    is_done: Mutex<bool>,
+    condvar: Condvar,
+}
+
+impl Done {
+    fn new() -> Self {
+        Self {
+            is_done: Mutex::new(false),
+            condvar: Condvar::new(),
+        }
+    }
+
+    fn wait_done(&self) {
+        let mut done = self.is_done.lock();
+        if !*done {
+            self.condvar.wait(&mut done)
```

Comment:
Something I'm not following here is, how does `done` get updated if `wait_done` holds the lock? :thinking: I can't piece it together from the `parking_lot` docs.

Reply:
By magic of RAII :rainbow: 

`done` is of type `MutexGuard`, but `Condvar` can unwrap the value. Essentially, it releases the lock internally, and parks the `wait`ing thread. Once another thread calls `notify_one`, the first waiting thread is unparked and acquires the lock again.

Reply:
:mage_man: